File: 2375-construct-smallest-number-from-di-string/main.go
func smallestNumber(pattern string) string {
    n := len(pattern)
    var ans []string
    var numStack []int
    
    for idx := range n + 1 {

        numStack = append(numStack, idx + 1)

        if idx == n || pattern[idx] == 'I' {
            for len(numStack) > 0 {
                top := numStack[len(numStack) - 1]
                numStack = numStack[:len(numStack) - 1]
                ans = append(ans, strconv.Itoa(top))
            }
        }
    }

    return strings.Join(ans, "")
}

File: 2375-construct-smallest-number-from-di-string/main.py
class Solution:
    def smallestNumber(self, pattern: str) -> str:
        ans = []
        num_stack = []
        n = len(pattern)

        for idx in range(n + 1):

            num_stack.append(idx + 1)

            if idx == n or pattern[idx] == "I":
                while num_stack:
                    ans.append(str(num_stack.pop()))
                    print(ans)

        return "".join(ans)
